fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,683 @@
+        argument 'first_input'. In this case the module 'first_input' must be added to the graph previously.
+        """
+        old_modules_len = len(self.modules_list)
+        new_modules_len = len(modules)
+        self._add_new_module(new_modules_len)
+        self.modules_list.extend(modules)
+        for i in range(old_modules_len + 1, old_modules_len + new_modules_len):
+            self.inputs[i] = [(i - 1, 0)]
+        if first_input is not None:
+            self.inputs[old_modules_len] = [(self.modules_list.index(first_input), 0)]
+
+    def feed_model_input(self, module: PipelineModule, ind: int = 0) -> None:
+        """Declares the input to a module as the input to the model. In case the model has multiple
+        inputs, the argument 'ind' indicates the index of the model input that is fed to the module.
+        """
+        self.inputs[self._find_or_add(module)] = [(-1, ind)]
+
+    def add_multi_input_layer(self, module: PipelineModule, inputs: List[PipelineModule]) -> None:
+        """Adds a module with multiple inputs to the graph. The modules that provide inpurs to this module
+        must be added previously to the graph and are listed with argument inputs.
+        """
+        self.inputs[self._find_or_add(module)] = [(self.modules_list.index(m), 0) for m in inputs]
+
+    def fan_out(self, module: PipelineModule, outputs: List[PipelineModule]) -> None:
+        """Feeds outputs of a previously added module to modules specified by argument 'outputs' (so
+        'module' should have at least 'len(outputs)' outputs.
+        Modules in the list 'outputs' are added to the graph if they have not been added previously.
+        """
+        mi = self.modules_list.index(module)
+        for i, m in enumerate(outputs):
+            self.inputs[self._find_or_add(m)] = [(mi, i)]
+
+    def replicate_output(self, module: PipelineModule, outputs: List[PipelineModule]) -> None:
+        """Feeds the first output of a previously added module to multiple modules specified by
+        argument 'outputs'. Modules in the list 'outputs' are added to the graph if they have not
+        been added previously.
+        """
+        mi = self.modules_list.index(module)
+        for m in outputs:
+            self.inputs[self._find_or_add(m)] = [(mi, 0)]
+
+    def validate_graph(self) -> None:
+        """Makes sure graph satisfies necessary requirements"""
+        # TODO: implement following checks:
+        #   * the graph has a least one module, and is connected.
+        #   * num_inputs and num_outputs for modules matche list of connections defined in the graph.
+        #   * all inputs to a module should come from model input, or modules with smaller index in
+        #     the graph. This condition is used in implementaion of DistributedPipeline.forward. Even
+        #     if we relax this condition, still need to make sure the graph is acyclic.
+        pass
+
+    def compute_output_users(self) -> None:
```

Comment:
Maybe rename this to `compile`. It was confusing to see a method named compute which doesn't return a value.